x-pack/legacy/plugins/security/server/lib/authentication/tokens.ts

```diff
@@ -0,0 +1,173 @@
+   */
+  public async invalidate({ accessToken, refreshToken }: Partial<TokenPair>) {
+    this.debug('Invalidating access/refresh token pair.');
+
+    let invalidationError;
+    let invalidatedTokensCount;
+    if (refreshToken) {
+      try {
+        invalidatedTokensCount = (await this.options.client.callWithInternalUser(
+          'shield.deleteAccessToken',
+          { body: { refresh_token: refreshToken } }
+        )).invalidated_tokens;
+      } catch (err) {
+        this.debug(`Failed to invalidate refresh token: ${err.message}`);
+        // We don't re-throw the error here to have a chance to invalidate access token if it's provided.
+        invalidationError = err;
+      }
+
+      if (invalidatedTokensCount === 0) {
+        this.debug('Refresh token was already invalidated.');
+      } else if (invalidatedTokensCount === 1) {
+        this.debug('Refresh token has been successfully invalidated.');
+      } else {
+        this.debug(
+          `${invalidatedTokensCount} refresh tokens were invalidated, this is unexpected.`
+        );
+      }
+    }
+
+    if (accessToken) {
+      try {
+        invalidatedTokensCount = (await this.options.client.callWithInternalUser(
+          'shield.deleteAccessToken',
+          { body: { token: accessToken } }
+        )).invalidated_tokens;
+      } catch (err) {
+        this.debug(`Failed to invalidate access token: ${err.message}`);
+        invalidationError = err;
+      }
+
+      if (invalidatedTokensCount === 0) {
+        this.debug('Access token was already invalidated.');
+      } else if (invalidatedTokensCount === 1) {
+        this.debug('Access token has been successfully invalidated.');
+      } else {
+        this.debug(`${invalidatedTokensCount} access tokens were invalidated, this is unexpected.`);
+      }
+    }
+
+    if (invalidationError) {
+      throw invalidationError;
+    }
+  }
+
+  /**
+   * Tries to determine whether specified error that occurred while trying to authenticate request
+   * using access token happened because access token is expired. We treat all `401 Unauthorized`
+   * as such. Another use case that we should temporarily support (until elastic/elasticsearch#38866
+   * is fixed) is when token document has been removed and ES responds with `500 Internal Server Error`.
+   * @param err Error returned from Elasticsearch.
+   */
+  public static isAccessTokenExpiredError(err?: any) {
+    const errorStatusCode = getErrorStatusCode(err);
+    return (
+      errorStatusCode === 401 ||
+      (errorStatusCode === 500 &&
+        !!(
+          err &&
+          err.body &&
+          err.body.error &&
+          err.body.error.reason === 'token document is missing and must be present'
+        ))
+    );
+  }
+
+  /**
+   * Logs message with `debug` level and tokens/security related tags.
+   * @param message Message to log.
+   */
+  private debug(message: string) {
+    this.options.log(['debug', 'security', 'tokens'], message);
```

Comment:
> Are we worried about losing the context of which auth provider is performing these token operations? 

Yeah, I thought about this, but I'm not really worried: when we see logs, these will be surrounded by something like this, so it's pretty easy to derive the correct context:
> [debug, security, saml] authenticating ......
> [debug, security, tokens] refreshing tokens.....
> [debug, security, saml] authentication completed.....

Having said that, our logs need some love and unifying :) Different providers were written by different people, and since there were no guidelines logs don't look as consistent as they should be.

Reply:
In NP `Logger` may provide factory method to extend parent context https://github.com/elastic/kibana/issues/39695
```js
// class SamlProvider
const log = log.get('saml');
new Tokens({log})
class Token {
  constructor({log: Logger}){
    this.log = log.get('tokens');
    this.log.debug('....'); // ] ['security', 'saml', 'tokens'] ....
  }
```